### ISSUE TITLE
Re-Strengthen declared derived ids invariant

### ISSUE THREAD

**Issue by cblappert** [facebook/prepack#2383]
Invariant was weakened in #2382, which exposed some issues in checking whether or not a derived id has been declared in the emitter/serializer:

Investigate: [this line](https://github.com/facebook/prepack/blob/393624f00f7ad5291622ee28bf90ea0470becbcf/src/serializer/Emitter.js#L103) and [this one](https://github.com/facebook/prepack/blob/29c36bafa52ba8bad154570d6a99fd545be9a099/src/serializer/ResidualHeapSerializer.js#L1949)

From the PR:

> I am not quite sure what we are trying to test with the last line of this invariant. I think what we're trying to test with the last line of this invariant is: If we're emitting to an additional function, it must be the case that referencedDeclaredValues has val. Here the test also passes when referencedDeclaredValues.has(val) is false which seems wrong to me.
> 
> I'm also dubious that top level optimized functions don't consider the MainGenerator as their parent generator, as this means that any derived ids declared in the MainGenerator shouldn't be considered declared in optimized functions (was able to repro that in conditions2 test after tightening invariants). For the sake of unblocking React Compiler, I'll look into that in a future PR.

**Comment by cblappert** [comment]
Logic has been changed so that issue is no longer relevant, closing.
